### PR TITLE
 Miri: use extern fn to expose interpreter operations to program; fix leak checker on Windows

### PR DESCRIPTION
**Pull request by RalfJung**
This PR realizes an idea that @oli-obk has been suggesting for a while: to use Miri-specific `extern` functions to provide some extra capabilities to the program. Initially, we have two of these methods, which libstd itself needs:
* `miri_start_panic`, which replaces the intrinsic of the same name (mostly for consistency, to avoid having multiple mechanisms for Miri-specific functionality).
* `miri_static_root`, which adds an allocation to a list of static "roots" that Miri considers as not having leaked (including all memory reachable through them). This is needed for https://github.com/rust-lang/miri/issues/1302.

We use `extern` functions instead of intrinsics for this so that user code can more easily call these Miri hoolks -- e.g. `miri_static_root` should be useful for https://github.com/rust-lang/miri/issues/1318.

The Miri side of this is at https://github.com/rust-lang/miri/pull/1485.

r? @oli-obk

### CODE REVIEW

File: src/libstd/sys/windows/thread_local_key.rs [hunk 1]
```diff
@@ -110,14 +110,29 @@ struct Node {
...
     let mut head = DTORS.load(SeqCst);
     loop {
         node.next = head;
         match DTORS.compare_exchange(head, &mut *node, SeqCst, SeqCst) {
-            Ok(_) => return mem::forget(node),
+            Ok(_) => {
+                #[cfg(miri)]
+                miri_static_root(&*node as *const _ as *const u8);
+
+                return mem::forget(node);
```
Comment (oli-obk) [inline src/libstd/sys/windows/thread_local_key.rs:134]: This is weird. I know it was pre-existing, but please drive-by fix it to return after invoking `mem::forget` instead of returning the unit result from that call?

**Comment by oli-obk** [comment]
I like how this change can be done without any boostrapping problems.

r=me with the nit

File: src/libstd/sys/windows/thread_local_key.rs (hunk 1 above)
Comment (RalfJung) [reply src/libstd/sys/windows/thread_local_key.rs:134]: Good point, I had not even realized this.

**Comment by RalfJung** [comment]
Fixed a typo.
@bors r=oli-obk
